doc/hardware/clock_management/index.rst

```diff
@@ -0,0 +1,1427 @@
+
+Clock Management Usage
+**********************
+
+In order to interact with the clock tree, clock consumers must define and
+initialize a clock output device. For devices defined in devicetree, using
+clocks defined within their ``clock-outputs`` property,
+:c:macro:`CLOCK_MANAGEMENT_DT_DEFINE_OUTPUT` or similar may be used. For
+software applications consuming a clock,
+:c:macro:`CLOCK_MANAGEMENT_DEFINE_OUTPUT` should be used.
+
+Clock consumers may then initialize their clock output device using
+:c:macro:`CLOCK_MANAGEMENT_DT_GET_OUTPUT` or similar for consumer devices
+defined in devicetree, or :c:macro:`CLOCK_MANAGEMENT_GET_OUTPUT` for software
+applications consuming a clock.
+
+Clock output devices can then be used with the clock management API to
+interface with the clock tree, as described below.
+
+Reading Clocks
+==============
+
+Once a diver has defined a clock output and initialized it, the clock rate can
+be read using :c:func:`clock_management_get_rate`. This will return the
+frequency of the clock output in Hz, or a negative error value if the clock
+could not be read.
+
+Consumers can also monitor a clock output rate. To do so, the application must
+first enable :kconfig:option:`CONFIG_CLOCK_MANAGEMENT_RUNTIME`. The application may
+then set a callback of type :c:type:`clock_management_callback_handler_t` using
+:c:func:`clock_management_set_callback`. One callback is supported per consumer.
+
+The clock management subsystem will issue a callback directly before applying a
+new clock rate, and directly after the new rate is applied. See
+:c:struct:`clock_management_event` for more details.
+
+Setting Clock States
+====================
+
+Each clock output defines a set of states. Clock consumers can set these states
+directly, using :c:func:`clock_management_apply_state`. States are described in
+devicetree, and are opaque to the driver/application code consuming the clock.
+
+Each clock consumer described in devicetree can set named clock states for each
+clock output. These states are described by the ``clock-state-n`` properties
+present on each consumer. The consumer can access states using macros like
+:c:macro:`CLOCK_MANAGEMENT_DT_GET_STATE`
+
+
+Requesting Clock Rates
+======================
+
+In some applications, the user may not want to directly configure clock nodes
+within their devicetree. The clock management subsystem allows applications to
+request a clock rate directly as well, by using
+:c:func:`clock_management_req_rate`.  If any states satisfy the frequency range
+request, the state offering the frequency closest to the maximum frequency
+requested. will be applied.  Otherwise if
+:kconfig:option:`CONFIG_CLOCK_MANAGEMENT_SET_RATE` is set, the clock management
+subsystem will perform runtime calculations to apply a rate within the requested
+range. If runtime rate calculation support is disabled, the request will fail if
+no defined states satisfy it.
```

Comment:
It's nice to see that both pre-defined states and flexible rate solving might be supported by the API